server/shared/src/main/scala/org/http4s/server/middleware/GZip.scala

```diff
@@ -34,12 +34,14 @@ object GZip extends GZipPlatform {
       bufferSize: Int = 32 * 1024,
       level: DeflateParams.Level = DeflateParams.Level.DEFAULT,
       isZippable: Response[G] => Boolean = defaultIsZippable[G](_: Response[G]),
+      isZipped: Request[G] => Boolean = defaultIsZipped[G](_: Request[G]),
```

Comment:
@matkob you are going to need to add an overload to not break bincompat, e.g. 

```scala

 def apply[F[_]: Functor, G[_]: Compression](
      http: Http[F, G],
      bufferSize: Int = 32 * 1024,
      level: DeflateParams.Level = DeflateParams.Level.DEFAULT,
      isZippable: Response[G] => Boolean = defaultIsZippable[G](_: Response[G]),
      isZipped: Request[G] => Boolean = defaultIsZipped[G](_: Request[G]),
  ): Http[F, G] = apply(http, bufferSize, level, isZippable, defaultIsZipped[G])

 def apply[F[_]: Functor, G[_]: Compression](
      http: Http[F, G],
      bufferSize: Int = 32 * 1024,
      level: DeflateParams.Level = DeflateParams.Level.DEFAULT,
      isZippable: Response[G] => Boolean = defaultIsZippable[G](_: Response[G]),
      isZipped: Request[G] => Boolean = defaultIsZipped[G](_: Request[G]),
  ): Http[F, G] = ...
```

Reply:
So actually, I am not sure if it makes sense to put "`Gzip`" and "`Gunzip`" together. Shouldn't they be exposed as different middlewares?

I know there's the issue on the client side, but this PR is about the server.

Reply:
@armanbilge Yeah, I think that makes sense.

Reply:
Coming from the user perspective here. I assumed that adding `GZip` middleware would cover everything related to gzip in my http4s server, so both incoming requests and outgoing responses. I was really surprised it did not.

In #4981 I saw a discussion about that but I'm not convinced there should be separate middlewares for decompressing and compressing, especially because this requires some further renaming and deprecations.

Reply:
> you are going to need to add an overload to not break bincompat

I wanted to add that overload but it didn't play well with its optional arguments - compiler was complaining about the ambiguity of the signatures. Also, after I removed this additional `isZipped` argument, bincompat shouldn't be an issue.

Reply:
> So for logging we have RequestLogger and ResponseLogger and then Logger which does both. Maybe we need something similar here?

Time will tell, I guess. And if there's need, anyone can contribute such enhanced granularity.